File: EntativaiOS/Managers/AdminManager.swift
import Foundation
import LocalAuthentication
import SwiftUI

// MARK: - Admin Manager (Founder Access Control)
class AdminManager: ObservableObject {
    static let shared = AdminManager()
    
    @Published var isAdminMode = false
    @Published var adminSessionExpiry: Date?
    
    private let sessionTimeout: TimeInterval = 15 * 60 // 15 minutes
    private var sessionTimer: Timer?
    
    private init() {}
    
    // MARK: - Founder Check
    
    /// Verifies if current user is the founder (@neoqiss)
    func isFounderAccount() -> Bool {
        guard let token = KeychainManager.shared.getToken() else {
            return false
        }
        
        // Decode JWT to check username and is_founder flag
        // For now, check username (in production, verify JWT payload)
        return checkFounderInToken(token)
    }
    
    private func checkFounderInToken(_ token: String) -> Bool {
        // Decode JWT payload
        let segments = token.components(separatedBy: ".")
        guard segments.count == 3 else { return false }
        
        // Base64 decode payload
        var base64 = segments[1]
        base64 = base64.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        
        while base64.count % 4 != 0 {
            base64.append("=")
        }
        
        guard let data = Data(base64Encoded: base64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        
        // Check username and is_founder flag
        let username = json["username"] as? String
        let isFounder = json["is_founder"] as? Bool ?? false
        
        return username == "neoqiss" && isFounder
    }
    
    // MARK: - Admin Panel Access
    
    /// Show admin panel with biometric authentication
    func showAdminPanel(from viewController: UIViewController? = nil) {
        guard isFounderAccount() else {
            print("⚠️ Not authorized for admin access")
            return
        }
        
        // Require biometric authentication
        authenticateWithBiometric(reason: "Access Admin Panel") { [weak self] success in
            if success {
                self?.presentAdminPanel(from: viewController)
            } else {
                self?.showError(message: "Biometric authentication failed")
            }
        }
    }
    
    private func presentAdminPanel(from viewController: UIViewController?) {
        DispatchQueue.main.async {
            self.isAdminMode = true
            self.startAdminSession()
            
            // Present admin panel
            let adminView = AdminPanelView()
            let hostingController = UIHostingController(rootView: adminView)
            hostingController.modalPresentationStyle = .fullScreen
            
            if let vc = viewController ?? UIApplication.topViewController() {
                vc.present(hostingController, animated: true)
            }
        }
    }
    
    // MARK: - Session Management
    
    private func startAdminSession() {
        adminSessionExpiry = Date().addingTimeInterval(sessionTimeout)
        
        // Invalidate existing timer
        sessionTimer?.invalidate()
        
        // Start new session timer
        sessionTimer = Timer.scheduledTimer(withTimeInterval: sessionTimeout, repeats: false) { [weak self] _ in
            self?.endAdminSession()
        }
    }
    
    func endAdminSession() {
        isAdminMode = false
        adminSessionExpiry = nil
        sessionTimer?.invalidate()
        sessionTimer = nil
    }
    
    func extendAdminSession() {
        guard isAdminMode else { return }
        
        authenticateWithBiometric(reason: "Extend Admin Session") { [weak self] success in
            if success {
                self?.startAdminSession()
            } else {
                self?.endAdminSession()
            }
        }
    }
    
    // MARK: - Quick Actions
    
    /// Ban a user with biometric confirmation
    func quickBanUser(userId: String, reason: String, completion: @escaping (Bool) -> Void) {
        authenticateWithBiometric(reason: "Ban User") { [weak self] success in
            guard success else {
                completion(false)
                return
            }
            
            self?.performBanUser(userId: userId, reason: reason, completion: completion)
        }
    }
    
    private func performBanUser(userId: String, reason: String, completion: @escaping (Bool) -> Void) {
        guard let token = KeychainManager.shared.getToken() else {
            completion(false)
            return
        }
        
        let url = URL(string: "\(APIConfig.adminBaseURL)/users/\(userId)/ban")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(getDeviceID(), forHTTPHeaderField: "X-Device-ID")
        
        let body = ["reason": reason, "duration": 0] // 0 = permanent
        request.httpBody = try? JSONEncoder().encode(body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                completion(error == nil)
            }
        }.resume()
    }
    
    /// Shadowban a user
    func quickShadowbanUser(userId: String, reason: String, completion: @escaping (Bool) -> Void) {
        authenticateWithBiometric(reason: "Shadowban User") { [weak self] success in
            guard success else {
                completion(false)
                return
            }
            
            self?.performShadowban(userId: userId, reason: reason, completion: completion)
        }
    }
    
    private func performShadowban(userId: String, reason: String, completion: @escaping (Bool) -> Void) {
        guard let token = KeychainManager.shared.getToken() else {
            completion(false)
            return
        }
        
        let url = URL(string: "\(APIConfig.adminBaseURL)/users/\(userId)/shadowban")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(getDeviceID(), forHTTPHeaderField: "X-Device-ID")
        
        let body = ["reason": reason]
        request.httpBody = try? JSONEncoder().encode(body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                completion(error == nil)
            }
        }.resume()
    }
    
    // MARK: - User Impersonation (Step-Up Authentication)
    
    /// Impersonate a user (requires password + biometric)
    func impersonateUser(userId: String, reason: String, password: String, completion: @escaping (Bool, String?) -> Void) {
        // Require detailed reason (min 20 chars)
        guard reason.count >= 20 else {
            completion(false, "Reason must be at least 20 characters")
            return
        }
        
        // First biometric check
        authenticateWithBiometric(reason: "Impersonate User - Step 1") { [weak self] success in
            guard success else {
                completion(false, "Biometric authentication failed")
                return
            }
            
            // Second authentication - password verification
            self?.verifyPasswordAndImpersonate(
                userId: userId,
                reason: reason,
                password: password,
                completion: completion
            )
        }
    }
    
    private func verifyPasswordAndImpersonate(
        userId: String,
        reason: String,
        password: String,
        completion: @escaping (Bool, String?) -> Void
    ) {
        guard let token = KeychainManager.shared.getToken() else {
            completion(false, "Not authenticated")
            return
        }
        
        let url = URL(string: "\(APIConfig.adminBaseURL)/users/\(userId)/impersonate")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(getDeviceID(), forHTTPHeaderField: "X-Device-ID")
        
        let body: [String: Any] = [
            "reason": reason,
            "password": password
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let impersonationToken = json["impersonation_token"] as? String else {
                DispatchQueue.main.async {
                    completion(false, "Failed to create impersonation session")
                }
                return
            }
            
            DispatchQueue.main.async {
                // Start 10-minute auto-termination timer
                self.startImpersonationTimer(userId: userId, token: impersonationToken)
                completion(true, impersonationToken)
            }
        }.resume()
    }
    
    private func startImpersonationTimer(userId: String, token: String) {
        // Auto-terminate impersonation after 10 minutes
        Timer.scheduledTimer(withTimeInterval: 10 * 60, repeats: false) { [weak self] _ in
            self?.endImpersonation(userId: userId, token: token)
        }
    }
    
    private func endImpersonation(userId: String, token: String) {
        guard let authToken = KeychainManager.shared.getToken() else { return }
        
        let url = URL(string: "\(APIConfig.adminBaseURL)/users/\(userId)/end-impersonation")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        request.setValue(getDeviceID(), forHTTPHeaderField: "X-Device-ID")
        
        URLSession.shared.dataTask(with: request).resume()
    }
    
    // MARK: - Helpers
    
    private func authenticateWithBiometric(reason: String, completion: @escaping (Bool) -> Void) {
        let context = LAContext()
        var error: NSError?
        
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            completion(false)
            return
        }
        
        context.evaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            localizedReason: reason
        ) { success, error in
            DispatchQueue.main.async {
                completion(success)
            }
        }
    }
    
    private func showError(message: String) {
        // Show error alert
        print("❌ Admin Error: \(message)")
    }
    
    private func getDeviceID() -> String {
        return UIDevice.current.identifierForVendor?.uuidString ?? "unknown"
    }
}

// MARK: - API Config Extension
extension APIConfig {
    static var adminBaseURL: String {
        return "http://localhost:8005/api/admin" // Admin service on port 8005
    }
}

// MARK: - UIApplication Extension (Helper)
extension UIApplication {
    static func topViewController(base: UIViewController? = nil) -> UIViewController? {
        let base = base ?? UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?.rootViewController
        
        if let nav = base as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }
        
        if let tab = base as? UITabBarController, let selected = tab.selectedViewController {
            return topViewController(base: selected)
        }
        
        if let presented = base?.presentedViewController {
            return topViewController(base: presented)
        }
        
        return base
    }
}

File: EntativaAndroid/app/src/main/kotlin/com/entativa/managers/AdminManager.kt
package com.entativa.managers

import android.content.Context
import android.util.Log
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * AdminManager - Founder access control for @neoqiss
 * 
 * Provides secure admin panel access with:
 * - Founder account verification
 * - Triple-tap gesture detection
 * - Biometric authentication
 * - Device whitelisting
 * - Session timeout (15 minutes)
 * - Full audit logging
 */
object AdminManager {
    private const val TAG = "AdminManager"
    private const val ADMIN_BASE_URL = "http://10.0.2.2:8005/api/admin" // Admin service
    private const val SESSION_TIMEOUT_MS = 15 * 60 * 1000L // 15 minutes
    
    private var isAdminMode = false
    private var sessionExpiryTime: Long = 0
    private var sessionTimer: android.os.CountDownTimer? = null
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    // MARK: - Founder Verification
    
    /**
     * Checks if current user is the founder (@neoqiss)
     */
    fun isFounderAccount(context: Context): Boolean {
        val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
        val token = prefs.getString("jwt_token", null) ?: return false
        
        // Decode JWT to check username and is_founder flag
        return checkFounderInToken(token)
    }
    
    private fun checkFounderInToken(token: String): Boolean {
        try {
            val parts = token.split(".")
            if (parts.size != 3) return false
            
            // Decode payload (base64url)
            val payload = parts[1]
            val decodedBytes = android.util.Base64.decode(
                payload.replace('-', '+').replace('_', '/'),
                android.util.Base64.URL_SAFE or android.util.Base64.NO_WRAP
            )
            
            val json = JSONObject(String(decodedBytes))
            val username = json.optString("username", "")
            val isFounder = json.optBoolean("is_founder", false)
            
            return username == "neoqiss" && isFounder
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decode token", e)
            return false
        }
    }
    
    // MARK: - Admin Panel Access
    
    /**
     * Shows admin panel with biometric authentication
     */
    fun showAdminPanel(activity: FragmentActivity) {
        if (!isFounderAccount(activity)) {
            Log.w(TAG, "Not authorized for admin access")
            return
        }
        
        // Require biometric authentication
        authenticateWithBiometric(
            activity = activity,
            title = "Admin Panel Access",
            subtitle = "Authenticate as Founder"
        ) { success ->
            if (success) {
                launchAdminPanel(activity)
            } else {
                // Show error
                Log.e(TAG, "Biometric authentication failed")
            }
        }
    }
    
    private fun launchAdminPanel(activity: FragmentActivity) {
        isAdminMode = true
        startAdminSession()
        
        // Launch admin panel activity
        // TODO: Create AdminPanelActivity
        Log.i(TAG, "Admin panel launched - Founder mode active")
    }
    
    // MARK: - Session Management
    
    private fun startAdminSession() {
        sessionExpiryTime = System.currentTimeMillis() + SESSION_TIMEOUT_MS
        
        // Cancel existing timer
        sessionTimer?.cancel()
        
        // Start new session timer
        sessionTimer = object : android.os.CountDownTimer(SESSION_TIMEOUT_MS, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                // Update session time
            }
            
            override fun onFinish() {
                endAdminSession()
            }
        }.start()
    }
    
    fun endAdminSession() {
        isAdminMode = false
        sessionExpiryTime = 0
        sessionTimer?.cancel()
        sessionTimer = null
        Log.i(TAG, "Admin session ended")
    }
    
    fun extendAdminSession(activity: FragmentActivity) {
        if (!isAdminMode) return
        
        authenticateWithBiometric(
            activity = activity,
            title = "Extend Admin Session",
            subtitle = "Re-authenticate to continue"
        ) { success ->
            if (success) {
                startAdminSession()
            } else {
                endAdminSession()
            }
        }
    }
    
    // MARK: - Quick Admin Actions
    
    /**
     * Ban a user with biometric confirmation
     */
    suspend fun quickBanUser(
        context: Context,
        activity: FragmentActivity,
        userId: String,
        reason: String
    ): Result<Unit> {
        return withContext(Dispatchers.Main) {
            var result: Result<Unit>? = null
            
            authenticateWithBiometric(
                activity = activity,
                title = "Ban User",
                subtitle = "Confirm action"
            ) { success ->
                if (success) {
                    kotlinx.coroutines.GlobalScope.launch(Dispatchers.IO) {
                        result = performBanUser(context, userId, reason)
                    }
                } else {
                    result = Result.failure(Exception("Authentication failed"))
                }
            }
            
            // Wait for result
            while (result == null) {
                kotlinx.coroutines.delay(100)
            }
            
            result!!
        }
    }
    
    private suspend fun performBanUser(
        context: Context,
        userId: String,
        reason: String
    ): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
            val token = prefs.getString("jwt_token", null) ?: throw Exception("Not authenticated")
            
            val json = JSONObject().apply {
                put("reason", reason)
                put("duration", 0) // 0 = permanent
            }
            
            val body = json.toString().toRequestBody("application/json".toMediaType())
            
            val request = Request.Builder()
                .url("$ADMIN_BASE_URL/users/$userId/ban")
                .post(body)
                .addHeader("Authorization", "Bearer $token")
                .addHeader("X-Device-ID", getDeviceID(context))
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(Exception("Ban failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to ban user", e)
            Result.failure(e)
        }
    }
    
    /**
     * Shadowban a user
     */
    suspend fun quickShadowbanUser(
        context: Context,
        activity: FragmentActivity,
        userId: String,
        reason: String
    ): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
            val token = prefs.getString("jwt_token", null) ?: throw Exception("Not authenticated")
            
            val json = JSONObject().apply {
                put("reason", reason)
            }
            
            val body = json.toString().toRequestBody("application/json".toMediaType())
            
            val request = Request.Builder()
                .url("$ADMIN_BASE_URL/users/$userId/shadowban")
                .post(body)
                .addHeader("Authorization", "Bearer $token")
                .addHeader("X-Device-ID", getDeviceID(context))
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(Exception("Shadowban failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to shadowban user", e)
            Result.failure(e)
        }
    }
    
    /**
     * Impersonate a user (step-up authentication with password)
     */
    suspend fun impersonateUser(
        context: Context,
        activity: FragmentActivity,
        userId: String,
        reason: String,
        password: String
    ): Result<String> {
        // Require detailed reason (min 20 chars)
        if (reason.length < 20) {
            return Result.failure(Exception("Reason must be at least 20 characters"))
        }
        
        return withContext(Dispatchers.Main) {
            var result: Result<String>? = null
            
            // First biometric check
            authenticateWithBiometric(
                activity = activity,
                title = "Impersonate User - Step 1",
                subtitle = "First authentication"
            ) { success ->
                if (success) {
                    // Second authentication with password
                    kotlinx.coroutines.GlobalScope.launch(Dispatchers.IO) {
                        result = performImpersonation(context, userId, reason, password)
                    }
                } else {
                    result = Result.failure(Exception("Biometric authentication failed"))
                }
            }
            
            // Wait for result
            while (result == null) {
                kotlinx.coroutines.delay(100)
            }
            
            result!!
        }
    }
    
    private suspend fun performImpersonation(
        context: Context,
        userId: String,
        reason: String,
        password: String
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
            val token = prefs.getString("jwt_token", null) ?: throw Exception("Not authenticated")
            
            val json = JSONObject().apply {
                put("reason", reason)
                put("password", password)
            }
            
            val body = json.toString().toRequestBody("application/json".toMediaType())
            
            val request = Request.Builder()
                .url("$ADMIN_BASE_URL/users/$userId/impersonate")
                .post(body)
                .addHeader("Authorization", "Bearer $token")
                .addHeader("X-Device-ID", getDeviceID(context))
                .build()
            
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()
            
            if (response.isSuccessful && responseBody != null) {
                val responseJson = JSONObject(responseBody)
                val impersonationToken = responseJson.optString("impersonation_token", "")
                
                if (impersonationToken.isNotEmpty()) {
                    // Start 10-minute auto-termination
                    startImpersonationTimer(context, userId, impersonationToken)
                    Result.success(impersonationToken)
                } else {
                    Result.failure(Exception("No impersonation token returned"))
                }
            } else {
                Result.failure(Exception("Impersonation failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to impersonate user", e)
            Result.failure(e)
        }
    }
    
    private fun startImpersonationTimer(context: Context, userId: String, token: String) {
        // Auto-terminate after 10 minutes
        object : android.os.CountDownTimer(10 * 60 * 1000L, 1000) {
            override fun onTick(millisUntilFinished: Long) {}
            
            override fun onFinish() {
                kotlinx.coroutines.GlobalScope.launch(Dispatchers.IO) {
                    endImpersonation(context, userId)
                }
            }
        }.start()
    }
    
    private suspend fun endImpersonation(context: Context, userId: String) = withContext(Dispatchers.IO) {
        try {
            val prefs = context.getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
            val token = prefs.getString("jwt_token", null) ?: return@withContext
            
            val request = Request.Builder()
                .url("$ADMIN_BASE_URL/users/$userId/end-impersonation")
                .post("{}".toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer $token")
                .addHeader("X-Device-ID", getDeviceID(context))
                .build()
            
            client.newCall(request).execute()
            Log.i(TAG, "Impersonation session ended")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to end impersonation", e)
        }
    }
    
    // MARK: - Biometric Authentication
    
    private fun authenticateWithBiometric(
        activity: FragmentActivity,
        title: String,
        subtitle: String,
        callback: (Boolean) -> Unit
    ) {
        val executor = ContextCompat.getMainExecutor(activity)
        
        val biometricPrompt = BiometricPrompt(
            activity,
            executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    callback(true)
                }
                
                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    callback(false)
                }
                
                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                    callback(false)
                }
            }
        )
        
        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setNegativeButtonText("Cancel")
            .build()
        
        biometricPrompt.authenticate(promptInfo)
    }
    
    // MARK: - Helpers
    
    private fun getDeviceID(context: Context): String {
        return android.provider.Settings.Secure.getString(
            context.contentResolver,
            android.provider.Settings.Secure.ANDROID_ID
        )
    }
}

/**
 * Multi-tap listener for gesture detection
 */
abstract class OnMultiTapListener(private val requiredTaps: Int) : android.view.View.OnClickListener {
    private var tapCount = 0
    private var lastTapTime = 0L
    private val tapTimeout = 500L // 500ms between taps
    
    override fun onClick(v: android.view.View?) {
        val currentTime = System.currentTimeMillis()
        
        if (currentTime - lastTapTime < tapTimeout) {
            tapCount++
            if (tapCount >= requiredTaps) {
                onMultiTap()
                tapCount = 0
            }
        } else {
            tapCount = 1
        }
        
        lastTapTime = currentTime
    }
    
    abstract fun onMultiTap()
}
